lib/kernel/src/application_controller.erl

```diff
@@ -1910,6 +2001,59 @@ scan_file(Str) ->
 	    {error, {none, load_file, "no ending <dot> found"}}
     end.
 
+read_fd_until_end_and_close(FileDescriptorId,
+                            TimeoutMilliseconds,
+                            MaxSizeBytes) ->
+    ReadLimit = 1024,
+    Read =
+        fun Read(Ref, _Fd, _Data, DataSize) when DataSize > MaxSizeBytes ->
+                Reason = io_lib:format("Max size ~w bytes exceeded",
+                                       [MaxSizeBytes]),
+                {Ref, error, Reason};
+            Read(Ref, Fd, Data, DataSize) ->
+                case file:read(Fd, ReadLimit) of
+                    eof ->
```

Comment:
What happens when we are at the end of the file descriptor but the OS or the parent program did not finish writing to it? Does it return `eof`? Does it block? Depending on the behavior, we can get nondeterministic behavior where the configuration will only succeed if whoever is writing it has written it quick enough.

In any case, the fd to file descriptor functionality is really neat! I can't wait to play with it as an alternative to the current port-based one, which buffers all input and can be problematic in some cases.

Reply:
Thanks for the question. Yes, `file:read/2` will block and does not return `eof` unless the the OS or the parent program has finished writing to it. However, we can still get the nondeterministic behavior that you describe as there is currently a timeout for reading a configuration from a file descriptor see [here](https://github.com/erlang/otp/pull/2873/files/099232f6de906f05664031b05ac1327a9bc1be2b#diff-2e7a405caf2dc0b5e567c6f01d62ad4b37ef9e517976ffc02c716900f2c87a9fR2046). This is meant as a feature to prevent `erl` from getting struck when reading a configuration from a file descriptor that never ends or is extremely slow. Currently, this timeout is set to two seconds.  If it times out, an error message describing what went wrong will be printed. Perhaps two seconds is too short? Do you think a timeout does more harm than good?

Reply:
I think a timeout is welcome but 2 seconds seems to be too low? Maybe someone is curling an internal endpoint or a service for this config and I can see it taking a bit more than that.I would go with 10 seconds or similar, enough to prevent hiccups but providing relatively soon feedback if something went wrong. But this is just gut feeling, so your guess is as good as mine. :)

Reply:
I have decided to log warnings if a file descriptor is slow instead of timing out (after discussing this with other members of the Ericsson Erlang/OTP team). This approach does not cause any problems for users with very slow file descriptors but can still give the users feedback if they pass broken file descriptors.

See: https://github.com/erlang/otp/pull/2873/commits/fffa1a550e658950d1b9f667873de1f257cc973a